File: jdk6u21_src/deploy/src/common/share/classes/com/sun/deploy/xml/XMLParser.java
/*
 * @(#)XMLParser.java	1.27 10/03/24
 *
 * Copyright (c) 2006, Oracle and/or its affiliates. All rights reserved.
 * ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 */
package com.sun.deploy.xml;

import java.io.IOException;
import com.sun.deploy.util.Trace;
import com.sun.deploy.util.TraceLevel;


public class XMLParser {
    public XMLNode _root;
    public String _source;
    public String _current;
    public int    _tokenType;
    public String _tokenData;
    public static final int TOKEN_EOF = 1;			//  null
    public static final int TOKEN_END_TAG = 2;			//  </
    public static final int TOKEN_BEGIN_TAG = 3;		//  <
    public static final int TOKEN_CLOSE_TAG = 4;		//  >
    public static final int TOKEN_EMPTY_CLOSE_TAG = 5;		//  />
    public static final int TOKEN_PCDATA = 6;			//  qwerty

    private BadTokenException _savedBte;
    /**
     * This array contains the predefined XML character entity references that
     * all XML parsers are required to recognize.
     */
    public static final GeneralEntity [] REQUIRED_CHARACTER_REFERENCES =
        {
            new GeneralEntity("quot", "\""),
            new GeneralEntity("amp",  "&"),
            new GeneralEntity("apos", "'"),
            new GeneralEntity("lt",   "<"),
            new GeneralEntity("gt",   ">")
        };

    /**
     * Construct an <code>XMLParser</code>.
     *
     * @param source  - the source text to parse.
     */
    public XMLParser(String source) {
	Trace.println("new XMLParser with source:", TraceLevel.TEMP);
	Trace.println(source, TraceLevel.TEMP);

        _source    = source;
        _current   = source;
        _root      = null;
        _tokenData = null;

        _savedBte = null;
    }

    public BadTokenException getSavedException() {
        return _savedBte;
    }

    public XMLNode parse() throws BadTokenException {
        try {
            nextToken(_current);
            _root = parseXMLElement();
	} catch (NullPointerException npe) {

	    Trace.println("NULL Pointer Exception: "+npe, TraceLevel.TEMP);
            throw npe;

        } catch (BadTokenException bte) {

	    Trace.println("JNLP Parse Exception: "+bte, TraceLevel.TEMP);
            throw bte;
	}

        if (Trace.isTraceLevelEnabled(TraceLevel.TEMP)) {
            Trace.println("\n\nreturning ROOT as follows:\n"+
        		_root, TraceLevel.TEMP);
        }
        return _root;
    }


    private void nextToken(String source) throws BadTokenException {

        _current = skipFilling(source);

        if (_current == null) {				// TOKEN_EOF
            _tokenType = TOKEN_EOF;
        } else if (_current.startsWith(CDStart)) {      // TOKEN_PCDATA
            _tokenType = TOKEN_PCDATA;
	    _current = skipPCData(source, '<');         // skip till next "<"
        } else if (_current.startsWith("</")) {		// TOKEN_END_TAG
            _tokenType = TOKEN_END_TAG;
            _current = skipXMLName(skipForward(_current,2,0));
        } else if (_current.startsWith("<")) {		// TOKEN_BEGIN_TAG
            _tokenType = TOKEN_BEGIN_TAG;
            _current = skipXMLName(skipForward(_current,1,0));
        } else if (_current.startsWith(">")) { 		// TOKEN_CLOSE_TAG
            _tokenType = TOKEN_CLOSE_TAG;
            _current = skipForward(_current,1,0);
        } else if (_current.startsWith("/>")) {		// TOKEN_EMPTY_CLOSE_TAG
            _tokenType = TOKEN_EMPTY_CLOSE_TAG;
            _current = skipForward(_current,2,0);
        } else {					// TOKEN_PCDATA
            _tokenType = TOKEN_PCDATA;
	    _current = skipPCData(source, '<');         // skip till next "<"
	}
    }

    private static final String CDStart = "<![CDATA[";
    private static final String CDEnd   = "]]>";


    /**
     * This method skips over any parsed character data (PCDATA).  It returns a
     * string that is a substring of the complete XML source, and begins with
     * the first character that is not part of the PCDATA.
     * <br />
     * PCDATA is the text within an element or an attribute value.  As it is
     * parsed, general entities are replaced by their replacement text. However,
     * any unparsed character data (CDATA) is left alone.
     * <br />
     * After this method completes, the parsed character data is stored in the
     * member variable <code>_tokenData</code>.
     * <br />
     * Note: This parser currently only recognizes the required character
     *       reference entities.  All other entities will be left as is.
     *
     * @param source       unparsed XML source that is assumed to start with
     *                     PCDATA.
     * @param delimitChar  character that marks the end of the current PCDATA
     *                     and the start of the next markup element.
     *
     * @return The remaining, unparsed, XML source that follows the PCDATA text.
     *
     * @throws BadTokenException  if the end of the parsed character data isn't
     *                            found, or if an incomplete CDATA element is
     *                            embedded inside the PCDATA.
     */
    private String skipPCData(String source, char delimitChar) throws BadTokenException
    {
        int    index  = source.indexOf(delimitChar);
        String retval = null;

        if (index >= 0)
        {
            // look for PCDATA
            int pcdi = source.indexOf(CDStart);
            if ((pcdi >= 0) && (pcdi <= index)) {
                // This CDATA also contains PCDATA.  The XML parser needs to
                // keep the text inside the PCDATA element as-is.  In other
                // words, it shouldn't do any entity replacement, or complain
                // if there are any characters that aren't normally allowed in
                // a CDATA element.
                String pre    = parseCharacterReferences(source.substring(0, pcdi));
                String remain = source.substring(pcdi + CDStart.length());
                int    end    = remain.indexOf(CDEnd);
                if (end >= 0) {
                    retval     = skipPCData(remain.substring(end + CDEnd.length()),
                                            delimitChar);
                    _tokenData = pre + remain.substring(0, end) + _tokenData;
                }
                else {
                    // Failed to find the end of the PCDATA element.  The stream
                    // is corrupt or the JNLP isn't valid XML.  Set _current to
                    // a point as close to the the problem as possible.
                    _current = remain;
                    throw (new BadTokenException("Found the start of a PCDATA element with no end marker.",
                                                 _source, getLineNumber()));
                }
            }
            else {
                // Everything up to the delimiter is CDATA, which needs to be
                // checked for general entities.  Any entities found should be
                // substituted with the entity replacement text.  Currently,
                // this parser only supports character reference entities.
                //
                // Everything after the delimiter is unparsed XML.
                _tokenData = parseCharacterReferences(source.substring(0, index));
                retval     = source.substring(index);
            }
        }
        else if (source.trim().length() != 0) {
            // Failed to find the end of the CDATA element.  The stream is
            // corrupt or the JNLP isn't valid XML.  Set _current to a point
            // as close to the the problem as possible.
            _current = source;
            throw (new BadTokenException("Failed to find the '" + delimitChar +
                                         "' charater that marks the end of a CDATA element.",
                                         _source, getLineNumber()));
        }

        return (retval);
    }

    private XMLNode parseXMLElement() throws BadTokenException {
        XMLNode self, child, sibling;
        XMLAttribute attribute, nextAttribute, firstAttribute;
        String name;
        int openTagLineNumber;

        if (_tokenType == TOKEN_BEGIN_TAG) {
            name = _tokenData;
            openTagLineNumber = getLineNumber();

            // Parse attributes. This section eats all input until
            // an EOF, a > or a />
            firstAttribute = parseXMLAttribute(_current);
            attribute = firstAttribute;
            while (attribute != null) {
                nextAttribute = parseXMLAttribute(_current);
		attribute.setNext(nextAttribute);
		attribute = nextAttribute;
            }

            // Create node for new element tag
            self =  new XMLNode(name, firstAttribute);


            // This will eihter be a TOKEN_EOF, TOKEN_CLOSE_TAG, or a
            // TOKEN_EMPTY_CLOSE_TAG
            nextToken(_current);
            if ((_tokenType != TOKEN_EMPTY_CLOSE_TAG) &&
                (_tokenType != TOKEN_CLOSE_TAG) &&
                (_tokenType != TOKEN_EOF)) {
                throw new BadTokenException(_source, getLineNumber());
            }

            if (_tokenType == TOKEN_EMPTY_CLOSE_TAG) {
                nextToken(_current);
                // We are done with the sublevel - fall through to
                // continue parsing tags at the same level */
            } else if (_tokenType == TOKEN_CLOSE_TAG) {
                nextToken(_current);

                // Parse until end tag if found
                child = parseXMLElement();
                if (child != null) {
                    self.setNested(child);
                    child.setParent(self);
                }

                if (_tokenType == TOKEN_END_TAG) {
                    String closeTokenName = _tokenData;
                    if (!name.equals(closeTokenName)) {
                        // set only the first exception in the bteThreadLocal
                        if (_savedBte == null) {
                            _savedBte = new BadTokenException("WARNING: <" + name + "> tag is not closed correctly",
                                    _source, openTagLineNumber);
                        }

                        Trace.println("<" + name + "> tag at line number " + openTagLineNumber +
                                " is not closed correctly", TraceLevel.TEMP);
                    }
                    // Find closing bracket '>' for end tag
                    do {
                        nextToken(_current);
                    } while ((_tokenType != TOKEN_EOF) &&
                            (_tokenType != TOKEN_CLOSE_TAG));
                    nextToken(_current);
                }
            }
            // Continue parsing rest on same level
            if (_tokenType != TOKEN_EOF) {
                // Parse rest of stream at same level
                sibling = parseXMLElement();
                self.setNext(sibling);
            }
            return self;

        } else if (_tokenType == TOKEN_PCDATA) {
            // Create node for pcdata
            self = new XMLNode(_tokenData);
            nextToken(_current);
            return self;
        }

	return null;

    }

    private XMLAttribute parseXMLAttribute(String source)
                         throws BadTokenException {
        if (source == null) {
            return null;
        }
        _current = skipFilling(source);

        if ((_current == null) || _current.startsWith(">") ||
            _current.startsWith("/>")) {
            return null;
        }
        // extract name
        _current = skipAttributeName(_current);
        String name = _tokenData;
        _current = skipFilling(_current);
        if (!_current.startsWith("=")) {
            // This is really an error. We ignore this, and just try
            // to parse an attribute out of the rest of the string
            if (source.equals(_current)) {
                // avoid infinate loop - move foward 1 ...
                _current = skipForward(_current, 1, 0);
            }
            return parseXMLAttribute(_current);
        }
        _current = skipForward(_current, 1, 0);
        _current = skipWhitespace(_current);

        String value;
        if ((_current.startsWith("\"")) || (_current.startsWith("\'"))) {
	    char quote = _current.charAt(0);
            _current = skipForward(_current, 1, 0);  // past quote
	    _current = skipPCData(_current, quote);  // past PCData
	    value = _tokenData;
            _current = skipForward(_current, 1, 0);  // past endquote
        } else {
            _current = skipNonSpace(_current);
            value = _tokenData;
        }
	if (value != null) {
	    value = value.trim();
	}
        return new XMLAttribute(name, value);
    }

    /**
     * Parses a string, replacing any recognized character reference entities
     * with the proper substitution values.
     * <br />
     * Currently, this parser only recognizes the 5 character reference entities
     * that all XML parsers are required to recognize (<code>&quot;</code>,
     * <code>&amp;</code>, <code>&apos;</code>, <code>&lt;</code>,
     * <code>&gt;</code>), plus character references to characters in the unicode
     * sequence in either decimal or hexadecimal format.
     * <br />
     * If a source string contains an unrecognized entity, then it is silently
     * ingnored, and left unchanged in the result.
     *
     * @param source  the source string to parse.
     *
     * @return The source string with all recognized character reference
     *         entities replaced with the appropriate replacement character.
     */
    private String parseCharacterReferences(String source)
    {
        String result = source;

        int start = source.indexOf("&");
        if (start >= 0) {
            String pre    = source.substring(0, start);
            String subst  = "&";    // be forgiving
            String remain = source.substring(start + 1);
            int    end    = source.indexOf(";", start);

            if (end > start) {
                // the character reference contains at least one character
                boolean known = false;

                subst  = source.substring(start + 1, end);
                remain = source.substring(end + 1);

                if (subst.startsWith("#")) {
                    try {
                        int     radix   = 10;
                        int     index   = 1;
                        char [] unicode = { '\0' };

                        if (subst.startsWith("#x"))
                        {
                            // this is a hexadecimal unicode value
                            radix = 16;
                            index = 2;
                        }
                        // no else required; this is a decimal unicode value

                        unicode[0] = (char) Integer.parseInt(subst.substring(index), radix);
                        subst      = new String(unicode);
                        known      = true;

                    }
                    catch (NumberFormatException nfe) {
                        // Don't need to do anything here.  Since this is not a
                        // known character reference entity, it will be preserved
                        // as is, and logged below.
                    }
                }
                else {
                    for (int i = 0; i < REQUIRED_CHARACTER_REFERENCES.length; i++) {
                        if (REQUIRED_CHARACTER_REFERENCES[i].equals(subst)) {
                            subst = REQUIRED_CHARACTER_REFERENCES[i].getValue();
                            known = true;
                            break;
                        }
                        // no else required; keep searching
                    }
                }

                if (known == false) {
                    subst = "&" + subst + ";";
                    Trace.println("Unrecognized character entity reference: " + subst,
                                  TraceLevel.BASIC);
                }
                // no else required; was a known entity
            }
            // no else required; this is an invalid character reference, but
            // don't let that stop further parsing

            result = pre + subst + parseCharacterReferences(remain);
        }
        // no else required; no character references

        return (result);
    }

    private String skipForward(String source, int index, int length) {
        if (index < 0 || (index+length) >= source.length()) return null;
        return source.substring(index+length);
    }

    private String skipNonSpace(String source) throws BadTokenException {
        int index = 0;
        if (source == null) {
            return null;
        }
        int length = source.length();
        while ((index < length) &&
               !Character.isWhitespace(source.charAt(index))) {
            index++;
        }
	return skipPCData(source, source.charAt(index));
    }

    private String skipWhitespace(String source) {
        int index = 0;
        if (source == null) {
            return null;
        }
        while ((index < source.length()) &&
	        Character.isWhitespace(source.charAt(index))) {
            ++index;
        }
        return source.substring(index);
    }

    private boolean legalTokenStartChar(char c) {
        return ((c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') ||
                (c == '_') || (c == ':' ));
    }

    private boolean legalTokenChar(char c) {
        return ((c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') ||
                (c >= '0' && c <= '9') ||
                (c == '_') || (c == ':' ) ||
                (c == '.') || (c == '-'));
    }

    private String skipAttributeName(String source) {

        if (source == null) {
            return null;
        }
        int index = source.indexOf("=");
        if (index >= 0) {
            _tokenData = source.substring(0, index);
	    if (_tokenData != null) {
		_tokenData = _tokenData.trim();
	    }
        } else {
            _tokenData = null;
        }
        return skipForward(source, index, 0);
    }

    private String skipXMLName(String source) {
        int index = 0;
        if (source == null) {
            return null;
        }
        if (legalTokenStartChar(source.charAt(0))) {
            index = 1;
            while ((index < source.length()) &&
                   legalTokenChar(source.charAt(index))) {
                index++;
            }
        }
        _tokenData = source.substring(0, index);
        if (_tokenData != null) {
	    _tokenData = _tokenData.trim();
        }
        return skipForward(source, index, 0);
    }

    private String skipXMLComment(String source) {
        if ((source != null) && (source.startsWith("<!--"))) {
            int index = source.indexOf("-->", 4);
            return skipForward(source, index, 3);
        }
        return source;
    }

    private String skipXMLDocType(String source) {
        // make sure we don't skip PCData
        if ((source != null) && (source.startsWith("<!")) &&
                source.startsWith(CDStart) == false) {
            int index = source.indexOf(">",2);
            return skipForward(source, index, 1);
        }
        return source;
    }

    private String skipXMLProlog(String source) {
        if ((source != null) && (source.startsWith("<?"))) {
            int index = source.indexOf("?>",2);
            return skipForward(source, index, 2);
        }
        return source;
    }

    private String skipFilling(String source) {
        String next, prev;
        next = source;
        do {
            prev = next;
            next = skipWhitespace(next);
            next = skipXMLComment(next);
            next = skipXMLDocType(next);
            next = skipXMLProlog(next);
        } while (next != prev);
        return next;
    }

    private int getLineNumber() {
        int end, lineCount;
        if (_current == null) {
            end = _source.length();
        } else {
            end = _source.indexOf(_current);
        }
        lineCount = 0;
        int index = 0;
        while ((index < end) && (index != -1)) {
            index = _source.indexOf("\n", index);
            if (index >= 0) {
		index++;
                lineCount++;
	    }
        }
        return lineCount;
    }
}

File: jdk6u21_src/deploy/src/javaws/share/native/xmlparser.c
/*
 * @(#)xmlparser.c	1.19 10/03/24
 * 
 * Copyright (c) 2006, Oracle and/or its affiliates. All rights reserved.
 * ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 */


#include "system.h"
#include "util.h"
#include "xmlparser.h"
#include "msgString.h"

#include <sys/stat.h>
#include <ctype.h>

/* Internal declarations */
static XMLNode*      ParseXMLElement(void);
static XMLAttribute* ParseXMLAttribute(void);
static char*         SkipWhiteSpace(char *p);
static char*         SkipXMLName(char *p);
static char*         SkipXMLComment(char *p);
static char*         SkipXMLDocType(char *p);
static char*         SkipXMLProlog(char *p);
static char*         SkipPCData(char *p);
static int           IsPCData(char *p);
static void          ConvertBuiltInEntities(char* p);
static void          SetToken(int type, char* start, char* end);
static void          GetNextToken(void);
static XMLNode*      CreateXMLNode(int type, char* name);
static XMLAttribute* CreateXMLAttribute(char *name, char* value);
static XMLNode*      ParseXMLElement(void);
static XMLAttribute* ParseXMLAttribute(void);
static void          FreeXMLAttribute(XMLAttribute* attr);
static void          PrintXMLAttributes(XMLAttribute* attr);
static void          indent(int indt);

/** Iterates through the null-terminated buffer (i.e., C string) and replaces all
 *  UTF-8 encoded character >255 with 255
 *
 *  UTF-8 encoding:
 *
 *   Range A:  0x0000 - 0x007F 
 *                               0 | bits 0 - 7
 *   Range B : 0x0080 - 0x07FF  : 
 *                               110 | bits 6 - 10
 *                               10  | bits 0 - 5
 *   Range C : 0x0800 - 0xFFFF  :        
 *                               1110 | bits 12-15
 *                               10   | bits  6-11
 *                               10   | bits  0-5
 */
static void RemoveNonAsciiUTF8FromBuffer(char *buf) {
    char* p;
    char* q;
    char c;
    p = q = buf;
    while(*p != '\0') {
        c = *p;
        if ( (c & 0x80) == 0) {
            /* Range A */
            *q++ = *p++;            
        } else if ((c & 0xE0) == 0xC0){
            /* Range B */
            *q++ = (char)0xFF;
            p++;
            if (*p) p++;
        } else {
            /* Range C */
            *q++ = (char)0xFF;
            p++;
            if (*p) p++;
            if (*p) p++;            
        }        
    }
    /* Null terminate string */
    *q = '\0';    
}

/* --------------------------------------------------------------------- */

static char* SkipWhiteSpace(char *p) {
    if (p != NULL) {       
        while(*p == 0x20 || *p == 0x09 || *p == 0x0d || *p == 0x0A) p++;
    }
    return p;
}

static char* SkipXMLName(char *p) {
    char c = *p;
    /* Check if start of token */
    if ( ('a' <= c && c <= 'z') ||
         ('A' <= c && c <= 'Z') || 
         c == '_' || c == ':') {
        
        while( ('a' <= c && c <= 'z') ||
               ('A' <= c && c <= 'Z') || 
               ('0' <= c && c <= '9') ||
               c == '_' || c == ':' || c == '.' || c == '-' ) {
            p++;
            c = *p;
	    if (c == '\0') break;
        }        
    }
    return p;
}

static char* SkipXMLComment(char *p) {
    if (p != NULL) {
        if (strncmp(p, "<!--", 4) == 0) {
            p += 4;
            do {
                if (strncmp(p, "-->", 3) == 0) return p+3;
                p++;
            } while(*p != '\0');         
        }
    }
    return p;
}

static char* SkipXMLDocType(char *p) {
    if (p != NULL) {
        if (strncmp(p, "<!", 2) == 0) {
            p += 2;
            while (*p != '\0') {
                if (*p == '>') {
                    return p+1;
                }
                p++;
            }
        }
    } 
    return p;
}

static char* SkipXMLProlog(char *p) {
    if (p != NULL) {
        if (strncmp(p, "<?", 2) == 0) {
            p += 2;
            do {
                if (strncmp(p, "?>", 2) == 0) return p+2;
                p++;
            } while(*p != '\0');            
        }
    }
    return p;
}

/* Search for the built-in XML entities:
 * &amp; (&), &lt; (<), &gt; (>), &apos; ('), and &quote(")
 * and convert them to a real character
 */
static void ConvertBuiltInEntities(char* p) {
    char* q;
    q = p;
    while(*p) {
      if (IsPCData(p)) {
	/* dont convert &xxx values within PData */
        char *end;
        end = SkipPCData(p);
        while(p < end) {
	    *q++ = *p++;
	}
      } else {
        if (strncmp(p, "&amp;", 5) == 0) {
            *q = '&';
            q++;
            p += 5;
        } else if (strncmp(p, "&lt;", 4)  == 0) {
            *q = '<';
            p += 4;
        } else if (strncmp(p, "&gt;", 4)  == 0) {
            *q = '>';
            p += 4;
        } else if (strncmp(p, "&apos;", 6)  == 0) {
            *q = '\'';
            p += 6;
        } else if (strncmp(p, "&quote;", 7)  == 0) {
            *q = '\"';
            p += 7;
        } else {
            *q++ = *p++;
        }
      }
    }
    *q = '\0';
}

/* ------------------------------------------------------------- */
/* XML tokenizer */

#define TOKEN_UNKNOWN             0
#define TOKEN_BEGIN_TAG           1  /* <tag */
#define TOKEN_END_TAG             2  /* </tag */
#define TOKEN_CLOSE_BRACKET       3  /* >  */
#define TOKEN_EMPTY_CLOSE_BRACKET 4  /* /> */
#define TOKEN_PCDATA              5  /* pcdata */
#define TOKEN_CDATA               6  /* cdata */
#define TOKEN_EOF                 7

static char* CurPos       = NULL;
static char* CurTokenName        = NULL;
static int   CurTokenType;
static int   MaxTokenSize = -1;

/* Copy token from buffer to Token variable */
static void SetToken(int type, char* start, char* end) {
    int len = end - start;
    if (len > MaxTokenSize) {
        if (CurTokenName != NULL) free(CurTokenName);
        CurTokenName = (char *)malloc(len + 1);
        MaxTokenSize = len;
    }
    
    CurTokenType = type;
    strncpy(CurTokenName, start, len);
    CurTokenName[len] = '\0';    
}

/* Skip XML comments, doctypes, and prolog tags */
static char* SkipFilling(void) {
    char *q = CurPos;

    /* Skip white space and comment sections */
    do {
        q = CurPos;        
        CurPos = SkipWhiteSpace(CurPos);
        CurPos = SkipXMLComment(CurPos); /* Must be called befor DocTypes */
        CurPos = SkipXMLDocType(CurPos); /* <! ... > directives */
        CurPos = SkipXMLProlog(CurPos);   /* <? ... ?> directives */
    } while(CurPos != q);

    return CurPos;
}


/* Parses next token and initializes the global token variables above 
   The tokennizer automatically skips comments (<!-- comment -->) and
   <! ... > directives.
*/
static void GetNextToken(void) {
    char *p, *q;

    /* Skip white space and comment sections */
    p = SkipFilling();

    if (p == NULL || *p == '\0') {
        CurTokenType = TOKEN_EOF;
        return;
    } else if (p[0] == '<' && p[1] == '/') {
        /* TOKEN_END_TAG */
        q = SkipXMLName(p + 2);        
        SetToken(TOKEN_END_TAG, p + 2, q);        
        p = q;
    } else  if (*p == '<') {
        /* TOKEN_BEGIN_TAG */
        q = SkipXMLName(p + 1);        
        SetToken(TOKEN_BEGIN_TAG, p + 1, q);        
        p = q;
    } else if (p[0] == '>') {
        CurTokenType = TOKEN_CLOSE_BRACKET;                
        p++;
    } else if (p[0] == '/' && p[1] == '>') {
        CurTokenType = TOKEN_EMPTY_CLOSE_BRACKET;        
        p += 2;
    } else {                
        /* Search for end of data */
        q = p + 1;
        while(*q && *q != '<') {
	    if (IsPCData(q)) {
		q = SkipPCData(q);
	    } else {
		q++;
	    }
	}
        SetToken(TOKEN_PCDATA, p, q);
        /* Convert all entities inside token */
        ConvertBuiltInEntities(CurTokenName);
        p = q;
    }
    /* Advance pointer to beginning of next token */
    CurPos = p;
}


static XMLNode* CreateXMLNode(int type, char* name) {
    XMLNode* node;
    node  = (XMLNode*)malloc(sizeof(XMLNode));
    node->_type = type;
    node->_name = name;
    node->_next = NULL;
    node->_sub  = NULL;
    node->_attributes = NULL;
    return node;
}


static XMLAttribute* CreateXMLAttribute(char *name, char* value) {
    XMLAttribute* attr;
    attr = (XMLAttribute*)malloc(sizeof(XMLAttribute));
    attr->_name = name;
    attr->_value = value;
    attr->_next =  NULL;
    return attr;
}


XMLNode* ParseXMLDocument(char* buf) {
    XMLNode* root;

    /* Remove UTF-8 encoding from buffer */
    RemoveNonAsciiUTF8FromBuffer(buf);

    /* Get first Token */
    CurPos = buf;
    GetNextToken();
    
    
    /* Parse document*/
    root =  ParseXMLElement();

    return root;
}

static XMLNode* ParseXMLElement(void) {
    XMLNode*  node     = NULL;
    XMLNode*  subnode  = NULL;
    XMLNode*  nextnode = NULL;
    XMLAttribute* attr = NULL;
    
    if (CurTokenType == TOKEN_BEGIN_TAG) { 

        /* Create node for new element tag */
        node = CreateXMLNode(xmlTagType, strdup(CurTokenName));
    
        /* Parse attributes. This section eats a all input until 
           EOF, a > or a /> */
        attr = ParseXMLAttribute();
        while(attr != NULL) {
          attr->_next = node->_attributes;
          node->_attributes = attr;
          attr = ParseXMLAttribute();
        }

        /* This will eihter be a TOKEN_EOF, TOKEN_CLOSE_BRACKET, or a 
         * TOKEN_EMPTY_CLOSE_BRACKET */
        GetNextToken();
        
        /* Skip until '>', '/>' or EOF. This should really be an error, */
        /* but we are loose */
        assert(CurTokenType == TOKEN_EMPTY_CLOSE_BRACKET ||
               CurTokenType == TOKEN_CLOSE_BRACKET ||
               CurTokenType  == TOKEN_EOF, getMsgString(MSG_XMLPARSE_ERROR));
        
        if (CurTokenType == TOKEN_EMPTY_CLOSE_BRACKET) {
            GetNextToken();
            /* We are done with the sublevel - fall through to continue */
            /* parsing tags at the same level */            
        } else if (CurTokenType == TOKEN_CLOSE_BRACKET) {
            GetNextToken();
            
            /* Parse until end tag if found */
            node->_sub  = ParseXMLElement();
            
            if (CurTokenType == TOKEN_END_TAG) {
                /* Find closing bracket '>' for end tag */
                do {
                   GetNextToken();                   
                } while(CurTokenType != TOKEN_EOF && CurTokenType != TOKEN_CLOSE_BRACKET);
                GetNextToken();
            }  
        }               
        
        /* Continue parsing rest on same level */
        if (CurTokenType != TOKEN_EOF) {
                /* Parse rest of stream at same level */
                node->_next = ParseXMLElement();
        }
        return node;

    } else if (CurTokenType == TOKEN_PCDATA) {
        /* Create node for pcdata */
        node = CreateXMLNode(xmlPCDataType, strdup(CurTokenName));
        GetNextToken();
        return node;
    }

    /* Something went wrong. */
    return NULL;
}

/* Parses an XML attribute. */
static XMLAttribute* ParseXMLAttribute(void) {
    XMLAttribute* attr = NULL;
    char* q;
    char* name;
    char* value;
    char quoteChar;

    /* Skip whitespace etc. */
    SkipFilling();

    /* Check if we are done witht this attribute section */
    if (CurPos[0] == '\0' ||
        CurPos[0] == '>' ||
        CurPos[0] == '/' && CurPos[1] == '>') return NULL;

    /* Find end of name */
    q = CurPos;
    while(*q && !iswspace(*q) && *q !='=') q++;

    SetToken(TOKEN_UNKNOWN, CurPos, q);
    name = strdup(CurTokenName);

    /* Skip any whitespace */
    CurPos = q;
    CurPos = SkipFilling();
       
    /* Next character must be '=' for a valid attribute */
    if (*CurPos != '=') {
       /* This is really an error. We ignore this, and just try to parse an attribute
          out of the rest of the string */
       return ParseXMLAttribute();            
    }   
        
    CurPos++;
    CurPos = SkipWhiteSpace(CurPos);
    /* Parse CDATA part of attribute */
    if ((*CurPos == '\"') || (*CurPos == '\'')) {
        quoteChar = *CurPos;
        q = ++CurPos;
        while(*q != '\0' && *q != quoteChar) q++;
        SetToken(TOKEN_CDATA, CurPos, q);
        CurPos = q + 1;
    } else {
        q = CurPos;
        while(*q != '\0' && !iswspace(*q)) q++;
        SetToken(TOKEN_CDATA, CurPos, q);
        CurPos = q;
    }        

    value = strdup(CurTokenName);
    attr  = CreateXMLAttribute(name, value);        
    
    return attr;
}


void FreeXMLDocument(XMLNode* root) {
  if (root == NULL) return;
  FreeXMLDocument(root->_sub);
  FreeXMLDocument(root->_next);
  FreeXMLAttribute(root->_attributes);
  free(root->_name);
  free(root);
}

static void FreeXMLAttribute(XMLAttribute* attr) {
  if (attr == NULL) return;
  free(attr->_name);
  free(attr->_value);
  FreeXMLAttribute(attr->_next);
  free(attr);
}

/* Find element at current level with a given name */
XMLNode* FindXMLChild(XMLNode* root, char* name) {
  if (root == NULL) return NULL;

  if (root->_type == xmlTagType && strcmp(root->_name, name) == 0) {
    return root;
  } 

  return FindXMLChild(root->_next, name);
}

/* Search for an attribute with the given name and returns the contents. Returns NULL if
 * attribute is not found
 */
char* FindXMLAttribute(XMLAttribute* attr, char* name) {
  if (attr == NULL) return NULL;
  if (strcmp(attr->_name, name) == 0) return attr->_value;
  return FindXMLAttribute(attr->_next, name);
}


void PrintXMLDocument(XMLNode* node, int indt) {
  if (node == NULL) return;

  if (node->_type == xmlTagType) {
     printf("\n");
     indent(indt);
     printf("<%s", node->_name);
     PrintXMLAttributes(node->_attributes);
     if (node->_sub == NULL) {      
       printf("/>\n");
     } else {
       printf(">", node->_name);
       PrintXMLDocument(node->_sub, indt + 1);
       indent(indt);
       printf("</%s>", node->_name);       
     }
  } else {          
    printf("%s", node->_name);
  }
  PrintXMLDocument(node->_next, indt);
}

static void PrintXMLAttributes(XMLAttribute* attr) {
  if (attr == NULL) return;

  printf(" %s=\"%s\"", attr->_name, attr->_value);
  PrintXMLAttributes(attr->_next);
}

static void indent(int indt) {
    int i;  
    for(i = 0; i < indt; i++) {
        printf("  "); 
    }
}

char *CDStart = "<![CDATA[";
char *CDEnd = "]]>";


static char* SkipPCData(char *p) {
    char *end = strstr(p, CDEnd);
    if (end != NULL) {
	return end+sizeof(CDEnd);
    }
    return (++p);
}

static int IsPCData(char *p) {
    return (strncmp(CDStart, p, sizeof(CDStart)) == 0);
}

